[clang-format] revert to string << string handling to previous default

mydeveloperday opened the pull request:
Fixes 88433

A change made to the handling of chevron operators causes a large amount of flux in code bases that were previously using clang-format, this fix reverts that change to the default behaviour but adds that new behaviour behind a new option.

mydeveloperday committed:
[clang-format] revery to string << string handling back to previous
default

Fixes 88433

A change made to the handling of chevron operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new behaviour
behind a new option.

mydeveloperday commented:
Fixes #88483

mydeveloperday commented:
Other related issues

#44363 , #59797

mydeveloperday committed:
[clang-format] revery to string << string handling back to previous default behaviour

mydeveloperday committed:
[clang-format] revery to string << string handling back to previous default

Fixes #88433

A change made to the handling of chevron operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option.

mydeveloperday committed:
[clang-format] revery to string << string handling to previous default

Fixes #88433

A change made to the handling of chevron operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option BreakChevronOperator.

mydeveloperday committed:
[clang-format] revert to string << string handling to previous default

Fixes #88433

A change made to the handling of chevron operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option BreakChevronOperator.

mydeveloperday committed:
[clang-format] revert to string << string handling to previous default

Fixes #88433

A change made to the handling of chevron operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option BreakChevronOperator.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5598,10 +5598,34 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
+    }
+  }
+  if (Style.BreakChevronOperator == FormatStyle::BCOS_Always) {
+    // can be std::os or os
```
I don't understand that comment.

mydeveloperday replied:
when you have 

os <<  "A" 
     << "B"

you don't want to break between os and << so the loop tracks back to determine if this is the first << in the line, this must also handle std::os << and a::b::os etc..  whats its really saying is don't break before the first <<.

HazardyKnusperkeks replied:
Yeah ok, but I'd skip the `os`.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5598,10 +5598,34 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
+  if (Style.BreakChevronOperator == FormatStyle::BCOS_Always) {
+    // can be std::os or os
+    auto *FirstChevron = Right.Previous;
+    while (FirstChevron) {
```
Is there a test case which utilizes this loop?

mydeveloperday replied:
yes anything with std::os in

mydeveloperday commented on clang/include/clang/Format/Format.h:
```diff
@@ -2193,6 +2193,41 @@ struct FormatStyle {
+
+  /// Break Between Chevron Operators
+  /// \version 19
+  BreakChevronOperatorStyle BreakChevronOperator;
```
So later commit in the PR do >> as well as << so maybe just BreakStreamOperator? WDYT

HazardyKnusperkeks replied:
That's good.

mydeveloperday committed:
[clang-format] revert to string << string handling to previous default

Fixes #88433

A change made to the handling of streaming operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option BreakStreamOperator.

mydeveloperday committed:
[clang-format] revert to string << string handling to previous default
Fixes #88433

A change made to the handling of streaming operators causes a large amount
of flux in code bases that were previously using clang-format, this fix
reverts that change to the default behaviour but adds that new
behaviours behind a new option BreakStreamOperator.

- Sort Alphabetically

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5598,11 +5598,45 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
+  }
+  case FormatStyle::BCOS_Always: {
+    // Don't break after the very first << or >>
+    // but the Left token can be os or std::os so
```
Still don't understand the part of `os` or `std::os`.

owenca commented:
Please see https://github.com/llvm/llvm-project/issues/88483#issuecomment-2053928993 and Conversation in #69859, which was the first attempt at adding an option for breaking consecutive stream insertion operations. Instead of handling everything specified in https://github.com/llvm/llvm-project/pull/69859#issuecomment-1776064714 in a single patch, we can take a few steps:

1. Handle when to break _before_ `<<`. The current default (since 18.x) is only when the left operand is a string literal ending with a newline and the right operand is a string literal.
2. Handle whether to indent or align after the break.
3. Do the above for breaking breaking _after_ `<<`.

(Ditto for stream extraction operations if needed.)

For No. 1 above, an `enum` option (e.g. `BreakBeforeStreamInsertionOperator`) with `Leave`, `AfterNewline` (LLVM default), `Always`, and `Never` may be sufficient. @s1Sharp
